myia/abstract/utils.py

```diff
@@ -167,6 +167,36 @@ def _unify(self, x: object, y: object, *, U):  # noqa: F811
         raise MapError(x, y, reason="Cannot merge objects")
 
 
+@ovld
+def _unify(  # noqa: F811
+    self,
+    x: data.AbstractUnion,
+    y: (data.AbstractAtom, data.AbstractStructure),
+    *,
+    U
+):
+    """Check if abstract value is in abstrct union."""
+    if y in x.options:
+        return y
+    else:
+        raise MapError(x, y, reason="Cannot merge objects")
```

Comment:
I think the type system would be more solid if it only allowed unifying a union with another union, so I'd like to check which tests require this to see if there's a better solution.